ticket_test.go

```diff
@@ -8,35 +8,81 @@ import (
 	"github.com/google/go-cmp/cmp/cmpopts"
 )
 
+var ignoreID = cmpopts.IgnoreFields(ticket.Ticket{}, "ID")
+
 func TestNewTicket(t *testing.T) {
 	t.Parallel()
-	p := ticket.NewProject("test")
-	want := ticket.Ticket{
-		Status:      ticket.StatusOpen,
-		Subject:     "My screen broke again",
-		Description: "Pixels missing!",
-	}
-	got := p.NewTicket(want.Subject)
-	if !cmp.Equal(want, got, cmpopts.IgnoreFields(want, "ID")) {
+	s := ticket.NewStore()
+	want := &ticket.Ticket{
+		Status:  ticket.StatusOpen,
+		Subject: "My screen broke again",
+	}
+	got := s.NewTicket(want.Subject)
+	if !cmp.Equal(want, got, ignoreID) {
 		t.Error(cmp.Diff(want, got))
 	}
 }
 
+func TestFields(t *testing.T) {
+	t.Parallel()
+	s := ticket.NewStore()
+	tk := s.NewTicket("testing something")
+	tk.Description = "Pixels missing"
+}
+
 func TestID(t *testing.T) {
 	t.Parallel()
-	p := ticket.NewProject("test")
-	t1 := p.NewTicket("test ticket")
+	s := ticket.NewStore()
+	t1 := s.NewTicket("test ticket")
 	if t1.ID == 0 {
 		t.Errorf("invalid id: %v", t1.ID)
 	}
-	t2 := p.NewTicket("another test ticket")
+	t2 := s.NewTicket("another test ticket")
+	// IDs are sequential - potential security issue in the future?
 	if t1.ID == t2.ID {
 		t.Errorf("want different IDs, got both == %v", t1.ID)
 	}
 }
 
+func TestGetTicket(t *testing.T) {
+	//create ticket in system
+	t.Parallel()
+	s := ticket.NewStore()
+
+	//pointer to created ticket
+	want := s.NewTicket("My screen broke again")
+
+	//look up ticket by ID and get a pointer to it
+	got, err := s.Get(want.ID)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	//both pointers should point to the same ticket
+	if !cmp.Equal(want, got) {
+		t.Error(cmp.Diff(want, got))
+	}
+}
+
+func TestGetByStatus(t *testing.T) {
+	s := ticket.NewStore()
+	tkOpen := s.NewTicket("this is open")
+	tkClosed := s.NewTicket("this will be closed")
+	tkClosed.Status = ticket.StatusClosed
+	want := []*ticket.Ticket{
+		tkOpen,
+	}
+	got, err := s.GetByStatus(ticket.StatusOpen)

```

Comment:
As we now know, this is a good start, but not sufficient. Perhaps we would have discovered the bug in GetByStatus() by manually changing `StatusOpen` here to `StatusClosed` and checking that the test failed—it wouldn't! But either way, we should now reinforce this test in such a way that it _demonstrates_ the bug, before fixing the function itself to _eliminate_ the bug (which the extended test will then prove has been fixed).

How could we do that? We have a ticket store with two tickets, one open, one closed. We have asked for all open tickets, and verified that the result matches a slice containing the open ticket. The logical extension to this is now to ask for all _closed_ tickets, and verify that the result matches a slice containing the closed ticket. We can add this code to the end of the existing test.

We now need a little more information in the failure messages, because the test can fail in two different cases. For example:

```go
t.Errorf("GetByStatus(StatusOpen): %v", cmp.Diff(want, got))
...
t.Errorf("GetByStatus(StatusClosed): %v", cmp.Diff(want, got))
```